libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteProgressEx.kt

```diff
@@ -8,35 +8,61 @@ import com.mapbox.api.directions.v5.models.RouteOptions
 import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
 import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.trip.model.RouteProgress
 import com.mapbox.navigation.base.utils.DecodeUtils.stepGeometryToPoints
+import com.mapbox.navigation.utils.internal.ifNonNull
 import com.mapbox.navigation.utils.internal.logD
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
 
 private const val LOG_CATEGORY = "NavigationRouteUtils"
 
 /**
- * This function checks whether the [NavigationRoute] has unexpected closures, which could be a reason to re-route.
+ * This function checks whether the [NavigationRoute] has unexpected upcoming closures, which could be a reason to re-route.
  *
- * `true` is returned whenever there's any [RouteLeg.closures] outside of a direct region around a coordinate from [RouteOptions.coordinatesList]
+ * The algorithm does not take to account current closures, where the puck is on.
+ *
+ * `true` is returned whenever there's any upcoming [RouteLeg.closures] outside of a direct region around a coordinate from [RouteOptions.coordinatesList]
  * that was allowed for snapping to closures with [RouteOptions.snappingIncludeClosuresList] or [RouteOptions.snappingIncludeStaticClosuresList].
  * flags. Otherwise, `false` is returned.
  */
 @ExperimentalPreviewMapboxNavigationAPI
-suspend fun NavigationRoute.hasUnexpectedClosures(): Boolean =
+suspend fun RouteProgress.hasUnexpectedUpcomingClosures(): Boolean =
     withContext(Dispatchers.Default) {
-        val snappingResultList = directionsRoute.getSnappingResultList()
+        val snappingResultList = navigationRoute.directionsRoute.getSnappingResultList()
+
+        val routeProgressData = ifNonNull(
+            currentLegProgress,
+            currentLegProgress?.currentStepProgress
+        ) { legProgress, stepProgress ->
+            RouteProgressData(legProgress.legIndex, stepProgress.stepIndex)
+        }
 
         var currentLegFirstWaypointIndex = 0
-        directionsRoute.legs()?.forEachIndexed { routeLegIndex, routeLeg ->
+        navigationRoute.directionsRoute.legs()?.forEachIndexed { routeLegIndex, routeLeg ->
 
             val legLastGeometryIndex by lazy {
-                directionsRoute.stepsGeometryToPoints(routeLeg).lastIndex
+                navigationRoute.directionsRoute.stepsGeometryToPoints(routeLeg).lastIndex
             }
 
             val silentWaypoints = routeLeg.silentWaypoints()
 
             routeLeg.closures()?.forEach { closure ->
+                if (routeProgressData != null) {
+                    if (routeProgressData.currentLegIndex > routeLegIndex) {
+                        // skipping passed closures
+                        return@forEach
+                    } else if (routeProgressData.currentLegIndex == routeLegIndex &&
+                        (
+                            routeProgressData.currentStepIndex >= closure.geometryIndexEnd() ||
```

Comment:
Does it make sense to compare `RouteStepProgress#stepIndex`, which is an index representing the current step the user is on, with `Closure#geometryIndexEnd`, which is the position in the coordinate list where the incident began, relative to the start of the leg it's on? They are indexed from different array, the first index from `RouteLeg.steps` while the latter from decoded route leg's geometry, aren't they? 🤔

Reply:
actually both indexes (`Closure#[anyIndex]` and `RouteStepProgress#stepIndex`) represent same index inside the leg. Could you clarify your concerns, if it's not what you mean?

Reply:
as i see in [the documentation](https://docs.mapbox.com/api/navigation/directions/#route-leg-object) closure index is "The position in the coordinate list where the closure began, relative to the start of the leg it's on.", so it sounds like it's related to geometry, not array of steps objects. Do you understand it in the same way?

Reply:
I see. They are the same: both represent the index inside the leg 👍 it also works in tests: they also rely on indexes inside the leg

Reply:
addressed the issue